### DISCUSSION TITLE
Electrostatic Shielding Model

### DISCUSSION THREAD

**Discussion by vermaprk** [idaholab/moose#25677] [Q&A General]
Hi
I am trying to model electric field shielding for a 2d domain as follow:
Initially the electric field E along x axis is a continuous variable and identical along y-axis (due to symmetry).
However, a material property `c(x,y,t)` is used for executing shielding condition. 
So whenever `c(x0,y0,t0) > c_cutoff` for x incrementing from x=0 to x = L (right boundary)
I want to execute E_new = E for x<=x0;
and,                      E_new = func(c(x0,y0,t0) ) for x>x0; irrespective of any other value of c along x.
Further I will be adding some smooth transitions for E.
Here E is a moose variable. 

In simple words for moving from x= 0 to x = xmax if we encounter c(x=x0)>c_cuttoff I want to define a unit function U(x-x0) irrespective of any other value of c(x>x0).
I am unable to find a way to execute this in moose. 

Please help.
Thanks

**Comment by GiudGiud** [comment]
Hello

So E is the variable that you are solving the equations for right?
We call this the 'nonlinear' variable.

There is no way to set a nonlinear variable, you can only solve equations to compute them.

So instead I suggest you make the terms of your equations dependent on c.
for example:

```
somekernel::computeQpResidual()
{
  if (c[_qp] < c_cutoff)
    normal term for the equation
  else
    different term that gets the value you want for E
}
```

alternatively, you can use the Control system to change between sets of kernels based on a criterion.
https://mooseframework.inl.gov/syntax/Controls/index.html

Guillaume

**Reply by vermaprk** [reply]
Yes, but its not just c[_qp] >= c_cutoff for the else case, rather its like a trigger. So as soon as c[qp]> c_cutoff, the else condition should always be true along the x axis for particular y. Will this code work in that case.
Thanks

**Reply by GiudGiud** [reply]
then you can keep a boolean called trigger in the kernel attributed
If the trigger happens, then set that boolean
and decide the term based on that boolean

The Controls system also handles triggers / persistent and non-persistent conditions

**Reply by vermaprk** [reply]
Couple of more thing 
1. how to reset the boolean once x=xmax for qp?
2. What is the order if loop will begin qp with? I mean is it like first for y0 it will go for all x0 to x1 and so on upto xn. Then shift to y1 and so on to yn for a 2 D system.

**Reply by GiudGiud** [reply]
is this a different condition for every qp?
if so a boolean on a per-kernel basis wont work btw. You ll need to create a field of booleans, either with an auxvariable and an auxkernel, or with material properties

you can have another if condition:
```
if (boolean and x > xmax)
  boolean = false
```

**Reply by vermaprk** [reply]
> is this a different condition for every qp?

The condition is same but once it is executed for all {x0,x1,x2.....xn} at y0. It needs to be reset and check again for the same condition to be true for y1 and .... .
My query is as every element has two qp. So will it begin with element 1 of y and look for the condition for x elements (starting from x1 to xn) to be true. Then reset the boolean at x=xmax. And then move to element 2 of y and repeat for all x elements.

**Reply by GiudGiud** [reply]
So there's a grid of elements? in X and Y?

We have layered user objects that can examine groups of elements in layers with an origin and an axis for layering

**Reply by vermaprk** [reply]
Yes I am using GeneratedMesh with nx and ny equal to 100.
Can't I just define a generic functor material that gives me a unit signal U(x-x0) if c(x0)>0. And then simply multiply it on my kernel.
Else, I will have to do it your way.
Thanks

**Reply by GiudGiud** [reply]
if you want to examine more than 1 element to make a decision on the value, you need something that works with layers

If you can decide based on the local value of c on one element, then yes, please go ahead and write an auxkernel or material property for the cirterion
